src/GitVersion.Core.Tests/Configuration/ConfigurationProviderTests.CanWriteOutEffectiveConfiguration.approved.txt

```diff
@@ -1,14 +1,27 @@
 assembly-versioning-scheme: MajorMinorPatch
 assembly-file-versioning-scheme: MajorMinorPatch
-mode: ContinuousDelivery
 tag-prefix: '[vV]'
-continuous-delivery-fallback-tag: ci
 major-version-bump-message: '\+semver:\s?(breaking|major)'
 minor-version-bump-message: '\+semver:\s?(feature|minor)'
 patch-version-bump-message: '\+semver:\s?(fix|patch)'
 no-bump-message: '\+semver:\s?(none|skip)'
 tag-pre-release-weight: 60000
-commit-message-incrementing: Enabled
+fallback-branch:
+  mode: ContinuousDelivery
+  tag: '{BranchName}'
+  increment: Inherit
+  prevent-increment-of-merged-branch-version: false
+  track-merge-target: false
+  commit-message-incrementing: Enabled
+  regex: ''
+  tracks-release-branches: false
+  is-release-branch: false
+  is-mainline: false
```

Comment:
If I think about it this might be also a possibility:
```csharp
public class GitVersionConfiguration : BranchConfiguration
{
    [YamlMember(Alias = "assembly-versioning-scheme")]
    public AssemblyVersioningScheme? AssemblyVersioningScheme { get; set; }

    [YamlMember(Alias = "assembly-file-versioning-scheme")]
    public AssemblyFileVersioningScheme? AssemblyFileVersioningScheme { get; set; }

    // ... and other general configuration properties

    [YamlMember(Alias = "unknown-branch")]
    public BranchConfiguration? UnknownBranch { get; set; }

    [YamlMember(Alias = "branches")]
    public Dictionary<string, BranchConfiguration> Branches { get; set; }
}
```
For the user it might be confusing what properties are branch related (fallback configuration) and what properties are used for the overall configuration. In my opinion it feels not intuitive.

What do you think?

Reply:
What do you think? @arturcic (see https://github.com/GitTools/GitVersion/pull/3235/#discussion_r1020707714 as well)

Reply:
I think the root configuration and the branch configuration could probably be mostly the same, but `branches` being an override for each branch category of a given repository. The root configuration should be the fallback configuration that applies to all branches until the configuration is overridden.

At least for most properties. Perhaps not all. It's something we need to consider and discuss.

Reply:
Okay no problem I'm going to move the fallback configuration section back to the root location.

I think this statement is not valid:
> The root configuration should be the fallback configuration that applies to all branches until the configuration is overridden.

At least the statement should be:

> The root configuration defines some fallback properties which applies on those property which are not specified on all branches but not marked with increment strategy `inherit`.

Please notice that we are not overriding the configuration from base (fallback configuration) to top (one or more branch configurations). We are inheriting from top (one or more branch configurations) to base (fallback configuration).

Reply:
Done

Reply:
Of course I agree we need the same terminology other wise it is confusing. With this discussion we are already building the ubiquitous language in the domain of GitVersion ;). That's good.

> What does "top" and "base" configuration refer to? Can we agree on common terms for these? I think the confusion and problem we have stems from the fact that the configuration for a given branch can come from two different and perhaps conflicting sources of truth:
> 
> 1. The parent branch (configured with increment: Inherit).
> 2. The root configuration.

In my opinion we have no root configuration for overriding the branch configuration. We have just branch configurations (if we ignore the unknown branch configuration for a moment) and one fallback configuration.

For what scenario do you need the root configuration? Like I already mentioned in this comment https://github.com/GitTools/GitVersion/issues/2388#issuecomment-1291713840 some users are expecting override all functionality which might be a use case. But this feature is just for convenience and gives no additional value to the product.

Reply:
If you take this example: pull-requests --inherits-from--> master --inherits-from--> fallback configuration and think about it how GitVersion works with the following configuration (the configuration after merging):

```yaml
mode: mainline
fallback:
  mode: ContinuousDelivery
  tag: '{BranchName}'
branches:
  main:
    mode: ContinuousDeployment
    increment: Patch
    regex: ^master$|^main$
    pre-release-weight: 55000
  pull-request:
    mode: ContinuousDelivery
    increment: Inherit
    regex: ^(pull|pull\-requests|pr)[/-]
    source-branches:
    - main
```

What is your expectation for the mode in the branch pull-requests?

Reply:
Okay, so what you call `fallback`, I'm thinking of as the "root" configuration. I don't think we need both. I also don't think `mode: null` should be supported. What we need is a way to clearly identify which relationship a property in the branch configuration should take its value from. The current `Inherit` is not sufficiently clear.

Your above example should imho look something like this.

```yaml
mode: ContinuousDelivery
branches:
  main:
    mode: ContinuousDeployment
    increment: Patch
    regex: ^master$|^main$
    pre-release-weight: 55000
  pull-request:
    mode: Default
    increment: Inherit
    regex: ^(pull|pull\-requests|pr)[/-]
    source-branches:
    - main
```

The `mode` for `pull-request` will be `ContinuousDelivery` since that is the `mode` of the root configuration. I'm not sure about the `Default` value, as it can be interpreted in one of two ways:

1. The default value of the `pull-request` branch category.
2. The value that is provided as the default in the root configuration.

So we need to iterate on this a bit to find a name that more clearly identifies the source of the value. Once we find a good name, I think we can say that all properties in a branch configuration has this value by default, except for the few properties that require a different value for the branch category (such as `mode: ContinuousDeployment` for `develop` and different values of `increment` for various categories).

Reply:
I also think we need to iterate on the name `Inherit`, since it does not clearly say that the value is inherited from the parent branch.

Reply:
Maybe we need to go one step back and discuss the user scenario. The discussion until now is quite technical and we need to focus one what user wants. I can think of a scenario where we have two user groups. One user group which uses ContinuousDelivery and another user group which uses Mainline mode. What we both want is that the user understands the configuration and is able to easily switch the mode.

Reply:
>I don't understand why you want to define the fallback configuration properties on the root configuration. If you doing this it's not clear for the user which properties are used as fallback and which properties are used for general purpose. 

Can't we define that the root configuration has no "general purpose" properties and that all properties on the root configuration provides a fallback for the branch configurations?

> That the fallback configuration section with all branch related properties (at least nine) needs to be placed on the root?

Yes (and thus "no" to your other two questions).

> One user group which uses ContinuousDelivery and another user group which uses Mainline mode. What we both want is that the user understands the configuration and is able to easily switch the mode.

Yes. For both of these examples, unless the users have specific requirements for how the different branch configurations need to differ from the defaults, the following configuration should be sufficient:

```yaml
mode: Mainline
```

And:

```yaml
mode: ContinuousDelivery
```